vapi/library/finder/finder.go

```diff
@@ -0,0 +1,309 @@
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package finder
+
+import (
+	"context"
+	"fmt"
+	"path"
+	"strings"
+
+	"github.com/vmware/govmomi/vapi/library"
+)
+
+// Finder is a helper object for finding content library objects by their
+// inventory paths: /LIBRARY/ITEM/FILE.
+//
+// Wildcard characters `*` and `?` are both supported. However, the use
+// of a wildcard character in the search string results in a full listing of
+// that part of the path's server-side items.
+//
+// Path parts that do not use wildcard characters rely on server-side Find
+// functions to find the path token by its name. Ironically finding one
+// item with a direct path takes longer than if a wildcard is used because
+// of the multiple round-trips. Direct paths will be more performant on
+// systems that have numerous items.
+type Finder struct {
+	M *library.Manager
+}
+
+// NewFinder returns a new Finder.
+func NewFinder(m *library.Manager) *Finder {
+	return &Finder{m}
+}
+
+// Find finds one or more items that match the provided inventory path(s).
+func (f *Finder) Find(
+	ctx context.Context, ipath ...string) ([]FindResult, error) {
+
+	if len(ipath) == 0 {
+		ipath = []string{""}
+	}
+	var result []FindResult
+	for _, p := range ipath {
+		results, err := f.find(ctx, p)
+		if err != nil {
+			return nil, err
+		}
+		result = append(result, results...)
+	}
+	return result, nil
+}
+
+func (f *Finder) find(ctx context.Context, ipath string) ([]FindResult, error) {
+
+	if ipath == "" {
+		ipath = "*"
+	}
+
+	// Get the argument and remove any leading separator characters.
+	if strings.HasPrefix(ipath, "/") {
+		ipath = ipath[1:]
+	}
+
+	// Tokenize the path into its distinct parts.
+	parts := strings.Split(ipath, "/")
+
+	libs, err := f.findLibraries(ctx, parts[0])
+	if err != nil {
+		return nil, err
+	}
+
+	// If the path is a single token then the libraries are requested.
+	if len(parts) < 2 {
+		return libs, nil
+	}
+
+	items, err := f.findLibraryItems(ctx, libs, parts[1])
+	if err != nil {
+		return nil, err
+	}
+
+	// If the path is two tokens then the library items are requested.
+	if len(parts) < 3 {
+		return items, nil
+	}
+
+	// Get the library item files.
+	return f.findLibraryItemFiles(ctx, items, parts[2])
+}
+
+// FindResult is the type of object returned from a Find operation.
+type FindResult interface {
+
+	// GetParent returns the parent of the find result. If the find result
+	// is a Library then this function will return nil.
+	GetParent() FindResult
```

Comment:
Doesn't matter too much, but: https://golang.org/doc/effective_go.html#Getters
If the `Get` prefix is go avoid clashing with field names, just leave as-is.

Reply:
Hi @dougm,

While the docs don't state so directly, I believe the suggestion is more applicable to `struct` types rather than `interface` types. Personally I think I should have made `FindResult` a `struct`, except originally it was an `interface` because there were more than one `findResult` `struct` variants.

Anyway, I made the functions have `Get` prefixes because as an interface I cannot know or assume that the structs that may implement the interface won't have conflicting, exported fields or functions. I thought `Get` prefixes would help avoid possible collisions.

Reply:
Hmm, now I remember the reason I didn't make `FindResult` a `struct`, even after removing any variants of the `findResult` type. Maybe I have it wrong, but I thought it is illegal to have a struct type reference itself in a field. However, I just tested that ([playground link](https://play.golang.org/p/dyh9dSZM8AX), and it doesn't appear to be the case.